ec/src/pairing.rs

```diff
@@ -0,0 +1,331 @@
+use ark_ff::{CyclotomicMultSubgroup, Field, One, PrimeField};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, SerializationError};
+use ark_std::{
+    borrow::Borrow,
+    fmt::{Debug, Display, Formatter, Result as FmtResult},
+    io::{Read, Write},
+    ops::{Add, AddAssign, Mul, MulAssign, Neg, Sub, SubAssign},
+    rand::{
+        distributions::{Distribution, Standard},
+        Rng,
+    },
+    vec::Vec,
+    UniformRand, Zero,
+};
+use zeroize::Zeroize;
+
+use crate::{AffineRepr, CurveGroup, Group};
+
+/// Collection of types (mainly fields and curves) that together describe
+/// how to compute a pairing over a pairing-friendly curve.
+pub trait Pairing: Sized + 'static + Copy + Debug + Sync + Send + Eq {
+    /// This is the scalar field of the G1/G2 groups.
+    type ScalarField: PrimeField;
+
+    /// An element in G1.
+    type G1: CurveGroup<ScalarField = Self::ScalarField, Affine = Self::G1Affine>
+        + From<Self::G1Affine>
+        + Into<Self::G1Affine>
+        // needed due to https://github.com/rust-lang/rust/issues/69640
+        + MulAssign<Self::ScalarField>;
+
+    type G1Affine: AffineRepr<Group = Self::G1, ScalarField = Self::ScalarField>
+        + From<Self::G1>
+        + Into<Self::G1>
+        + Into<Self::G1Prepared>;
+
+    /// A G1 element that has been preprocessed for use in a pairing.
+    type G1Prepared: Default
+        + Clone
+        + Send
+        + Sync
+        + Debug
+        + for<'a> From<&'a Self::G1>
+        + for<'a> From<&'a Self::G1Affine>
+        + From<Self::G1>
+        + From<Self::G1Affine>;
+
+    /// An element of G2.
+    type G2: CurveGroup<ScalarField = Self::ScalarField, Affine = Self::G2Affine>
+        + From<Self::G2Affine>
+        + Into<Self::G2Affine>
+        // needed due to https://github.com/rust-lang/rust/issues/69640
+        + MulAssign<Self::ScalarField>;
+
+    /// The affine representation of an element in G2.
+    type G2Affine: AffineRepr<Group = Self::G2, ScalarField = Self::ScalarField>
+        + From<Self::G2>
+        + Into<Self::G2>
+        + Into<Self::G2Prepared>;
+
+    /// A G2 element that has been preprocessed for use in a pairing.
+    type G2Prepared: Default
+        + Clone
+        + Send
+        + Sync
+        + Debug
+        + for<'a> From<&'a Self::G2>
+        + for<'a> From<&'a Self::G2Affine>
+        + From<Self::G2>
+        + From<Self::G2Affine>;
+
+    /// The extension field that hosts the target group of the pairing.
+    type TargetField: CyclotomicMultSubgroup;
+
+    /// Computes the product of Miller loops for some number of (G1, G2) pairs.
+    #[must_use]
+    fn multi_miller_loop(
```

Comment:
I think I understand the reason for naming it `multi_miller_loop` (base & parallel configuration squashed into one?), but it could be confusing - it almost sounds like multiple complete runs of Miller loop are computed. Why not stick to the old `miller_loop`?

Reply:
I think it is because the implementation here does not strictly follow what a Miller loop should be.

We often say Miller loop as f_{T, P)(Q) or its variants. In this case, we are actually doing the following, rather than a common Miller loop.
f_{T, P_1}(Q_1) \cdot f_{T, P_2}(Q_2) \cdot ...

The difference will be evident when one does a large number of Miller loops, likely in the case of **batch verification of zero-knowledge proofs** in order to improve the TPS, because of the parallelization support here.

Reply:
we could also call this `batch_miller_loop` and the corresponding pairing method `batch_pairing`. Does that sound better?

Reply:
Batching usually refers to **checking** a bunch, likely involving a randomizer. In this Miller loop we just add them together.